### PR TITLE
Convert mark-for-deployment to a state machine managed by pytransitions

### PR DESCRIPTION
**Pull request by EvanKrall**

### CODE REVIEW

File: paasta_tools/automatic_rollbacks.py [hunk 1]
```diff
@@ -110,3 +118,81 @@ def watch_for_slack_webhooks():
...
+            model=self,
+            states=list(self.states()),
+            transitions=list(self.valid_transitions()),
+            initial=self.start_state(),
+            after_state_change=self.after_state_change,
+            queued=True,
+        )
+
+        for state in list(self.states()):
+            # Call e.g. on_enter_start(self.enter_start)
```
Comment (solarkennedy) [inline paasta_tools/automatic_rollbacks.py]: Why do we need this? The docs say this is automatic?

File: paasta_tools/cli/cmds/mark_for_deployment.py [hunk 2]
```diff
@@ -423,60 +427,21 @@ def paasta_mark_for_deployment(args):
...
-    slack_notifier = SlackDeployNotifier(
-        deploy_info=deploy_info, service=service,
-        deploy_group=deploy_group, commit=commit, old_commit=old_git_sha, git_url=args.git_url,
-        auto_rollback=args.auto_rollback,
-    )
 
-    ret = mark_for_deployment(
-        git_url=args.git_url,
-        deploy_group=deploy_group,
+    deploy_process = MarkForDeploymentProcess(
```
Comment (solarkennedy) [inline paasta_tools/cli/cmds/mark_for_deployment.py]: Please gate this on `if args.auto_rollback`.

I know it is cleaner, but we have no itests, and this code can stand to wait until it has proven itself at least a little bit before we put it on the critical path of deploying all services at yelp.

File: paasta_tools/automatic_rollbacks.py (hunk 1 above)
Comment (EvanKrall) [reply paasta_tools/automatic_rollbacks.py]: Oh, I missed the bit where `on_enter_*` on your model gets automatically called; I only saw the part where you can pass a function/method name to `on_enter_*` on your Machine.

File: paasta_tools/cli/cmds/mark_for_deployment.py [hunk 3]
```diff
@@ -783,7 +999,8 @@ def wait_for_deployment(service, deploy_group, git_sha, soa_dir, timeout):
                  .format(git_sha, deploy_group))
 
     deadline = time.time() + timeout
-    green_light = Event()
+    if green_light is None:
+        green_light = Event()
```
Comment (solarkennedy) [inline paasta_tools/cli/cmds/mark_for_deployment.py]: Is this safe to do even if there are not threads going on?

Comment (EvanKrall) [reply paasta_tools/cli/cmds/mark_for_deployment.py]: Yep, pretty sure.

**Comment by keymone** [comment]
revert #2168 to apply this again

**Pull request merged**